test_conformance/compiler/test_build_helpers.cpp

```diff
@@ -550,6 +578,128 @@ REGISTER_TEST(get_program_info_kernel_names)
+        error =
+            clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, kernel_names_len,
+                             kernel_names.data(), &kernel_names_len);
+        test_error(error, "Unable to get kernel names list.");
+        std::string program_names = kernel_names.data();
+
+        std::vector<std::string> expected_names = { "sample_test_A",
+                                                    "sample_test_C" };
+        for (const auto &name : expected_names)
+        {
+            test_assert_error(program_names.find(name) != std::string::npos,
+                              "Unexpected kernel name");
+        }
+
+        std::vector<std::string> unexpected_names = { "sample_test_B",
+                                                      "sample_test_D" };
+        for (const auto &name : unexpected_names)
+        {
+            test_assert_error(program_names.find(name) == std::string::npos,
+                              "Unexpected kernel name");
+        }
+    }
+
+    // 2) Compile and link the programs with the preprocessor macro defined.
+    //    Query CL_PROGRAM_NUM_KERNELS and check that the correct number is
+    //    returned. Query CL_PROGRAM_KERNEL_NAMES and check that the right
+    //    kernel names are returned.
+    {
+        const char *build_options_B = "-DUSE_SAMPLE_TEST_B";
+        error = clCompileProgram(program_AB, 1, &device, build_options_B, 0, 0,
+                                 0, 0, 0);
+        test_error(error, "clCompileProgram failed");
+
+        const char *build_options_D = "-DUSE_SAMPLE_TEST_D";
+        error = clCompileProgram(program_CD, 1, &device, build_options_D, 0, 0,
+                                 0, 0, 0);
+        test_error(error, "clCompileProgram failed");
+
+        cl_program progs[] = { program_AB, program_CD };
+        program =
+            clLinkProgram(context, 1, &device, "", 2, progs, 0, 0, &error);
+        test_error(error, "clLinkProgram failed");
+
+        error = clGetProgramInfo(program, CL_PROGRAM_NUM_KERNELS,
+                                 sizeof(size_t), &total_kernels, nullptr);
+        test_error(error, "clGetProgramInfo failed");
+
+        test_assert_error(total_kernels == 4,
+                          "Unexpected clGetProgramInfo result");
+
+        error = clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, 0, nullptr,
+                                 &kernel_names_len);
+        test_error(error, "clGetProgramInfo failed");
+
+        std::vector<std::string> expected_names = {
+            "sample_test_A", "sample_test_B", "sample_test_C", "sample_test_D"
+        };
+
+        const size_t len = kernel_names_len + 1;
+        std::vector<char> kernel_names(len, '\0');
+        error =
+            clGetProgramInfo(program, CL_PROGRAM_KERNEL_NAMES, kernel_names_len,
+                             kernel_names.data(), &kernel_names_len);
+        test_error(error, "Unable to get kernel names list.");
```

Comment:
Minor: Could you change this to "Could not find expected kernel name" to better reflect the operation"

Reply:
Done